### PR TITLE
Add a descripiton and examples of key representation.

### LINKED ISSUE #46: Key Representation
... based on Ori's PR:
https://github.com/transmute-industries/cose/pull/8/files

### PR DESCRIPTION
**Pull request by dajiaji**
@hannestschofenig @OR13 @laurencelundblade

I have written a description and examples of key representation for COSE-HPKE.

Could you please review it to see if it matches your assumption?

Closes #46

### CODE REVIEW

**Review by OR13 — APPROVED** [review]
Great first start! I'm not sure about key ops in COSE key, I'm more familiar with it in JWK... I'd want to double check the empty recommendation vs omiting it.

**Comment by ilaril** [comment]
I think that key ops stuff is not correct. Looks like empty key_ops makes the key unusable, and DeriveBits is very strange operation for  decryption (since it is explicitly not for keys).

Here is what I think, from reading RFC 9052 (but I am not sure this is correct):
* For one-layer structure, the correct key operations are 3/4 (Encrypt/Decrypt).
* For two-layer structure, the correct key operations are 5/6 (Wrap/Unwrap).
* There should be no restrictions on what key ops are allowed beyond  RFC 9052.
  * Some combinations are not a good idea, but that is BCP stuff.

**Comment by dajiaji** [comment]
@ilaril Thanks for your comment.

> I think that key ops stuff is not correct.  

First of all, although the alg value is defined as the conbination of KEM/KDF/AEAD, the alg value for COSE_Key is optional and I think the key_ops parameter is only for the KEM keys. Unless considering this way, I think we cannot keep the consistency with the previous definition of the use of key_ops below:

The section 6.3.1 (Direct ECDH) says that the key_ops must be used as follows: 

> If the "key_ops" field is present, it MUST include "derive key" or "derive bits" for the private key.
> If the "key_ops" field is present, it MUST be empty for the public key.

In the case of the section 6.4 (ECDH with Key Wrap):

> If the "key_ops" field is present, it MUST include "derive key" or "derive bits" for the private key.
> If the "key_ops" field is present, it MUST be empty for the public key.

In addition, unless defining key_ops in this way, the key cannot be available on the Web Cryptography API. It will be considered an invalid key (this is the case of JWK but we have to consier the consistency with JWK definition).

> Looks like empty key_ops makes the key unusable,

Actually, my implementation also has a bug in the processing of key_ops (I have to fix it as soon as possible), but are there any other examples that the keys that have empty key_ops  cannot be used?

> and DeriveBits is very strange operation for decryption (since it is explicitly not for keys).

I agree that deriveBits is a strange thing.

However, since the Web Cryptography API used to implement a HPKE KEM operation is `deriveBits`, not `deriveKey`, I think we should limit the use of key_ops to `deriveBits`. In fact, the shared_secret generated as a result of KEM is not a key, just keying material.

**Comment by laurencelundblade** [comment]
The COSE HPKE draft may not define identifiers and uses for PQ algorithms, but it will allow them in the future, right? That means the key type might not be EC2 or OKP, so you can't just say the key type must be EC2 or OKP.  When RFC 9053 says that, it is for specific algorithms.

We need to do similar to what 9053 does. Something like listing the ciphersuites for which a particular definition of a COSE_key applies.

**Comment by ilaril** [comment]
Ugh, this key_ops stuff looks like a total mess. The webcrypto key operations just are not expressive enough:

* There are no ECDH/KEM public key operations.
* There is no direct encryption operation (like one-layer HPKE).
* However, there are "Key Transport Encryption" and "Key Wrap Encryption", which are public key operations (as well as "Verify"). What is the difference? Are any algorithms capable of both?

I wonder if the best way out would be to do whatever dumb thing WebCrypto requires (with disclaimers that this was due to shortcomings of WebCrypto) and define key_ops2 with actually sane  semantics.

Regarding key types, the acceptable key (sub)types should be listed by HPKE KEM:
* KEM 16/19 use EC2/1
* KEM 17/20 use EC2/2
* KEM 18/21 use EC2/3
* KEM 22 uses EC2/8
* KEM 32 uses OKP/4
* KEM 33 uses OKP/5
* KEM 48 uses OKP/TBD (needs to be defined)

(Of course, the list should be restricted to entries that are actually used.)

All the entries happen to use EC2 or OKP, and this will probably continue to hold, but there are no guarantees it will continue to hold. Some KEMs might even be able to use multiple key subtypes.

**Comment by laurencelundblade** [comment]
On Nov 9, 2023, at 12:05 PM, ilaril ***@***.******@***.***>> wrote:


Regarding key types, the acceptable key (sub)types should be listed by HPKE KEM


No and yes.

No because we are using ciphersuites now, not KEM IDs.

Yes, because they vary by KEM and the COSE-HPKE draft is for current AND future KEMS

LL

**Comment by ilaril** [comment]
Each cipher suite maps to one KEM (signature * -> 1), and each KEM maps to key subtypes (signature 1 -> *).

So specifying key subtypes for cipher suites directly would be denormal.

**Comment by laurencelundblade** [comment]
Agreed that it denormal and we could specify by KEM with some text that says there is mapping to the ciphersuite.

But denormal is not a big deal here because there aren't that many and they will be added rarely so it could be by ciphersuite too.

**Comment by dajiaji** [comment]
> The COSE HPKE draft may not define identifiers and uses for PQ algorithms, but it will allow them in the future, right? That means the key type might not be EC2 or OKP, so you can't just say the key type must be EC2 or OKP. When RFC 9053 says that, it is for specific algorithms.

I also anticipate adding key types for post-quantum hybrid KEM in the future, and I have taken care not to limit it to just EC2 and OKP. Would the following description be insufficient?

```
* The "kty" field MUST be present, and it MUST be one of the key types for HPKE KEM.
* If the "kty" field is "OKP" or "EC2", the "crv" field MUST be present
  and it MUST be a curve for HPKE KEM.
```
I intended to include future-defined KEMs in the statement "the key types for HPKE KEM.”

> We need to do similar to what 9053 does. Something like listing the ciphersuites for which a particular definition of a COSE_key applies.

Okay. I'll add it.

**Comment by dajiaji** [comment]
> I wonder if the best way out would be to do whatever dumb thing WebCrypto requires (with disclaimers that this was due to shortcomings of WebCrypto) and define key_ops2 with actually sane semantics.

I agree with your thoughts on key_ops to some extent. However, regarding the key_ops, I'd like to proceed with this PR as it is. 

> KEM 48 uses OKP/TBD (needs to be defined) 

Hmm, I don't think OKP can be used for KEM48. I think we should define an additional key type for Hybrid KEM, which looks like `HPKE-KEM` that I proposed before.

**Comment by ilaril** [comment]
Well, key_ops2 stuff would be different draft, just leaving note that the weird stuff is because of WebCrypto compatibility.

OKP can be used for KEM48 specifically, and should be, because it is the matching keyshape.

HPKE-KEM key type would have been very useful with a'la carte stuff, but it is less useful now with cipher suites.

**Comment by dajiaji** [comment]
@ilaril @laurencelundblade 

I have added a mapping table of alg, kty and crv. Could you check whether it's in line of your thoughts?

**Comment by OR13** [comment]
Let's avoid signatures for this draft, there is work at cfrg that is still being worked out for hybrid signatures.

**Comment by selfissued** [comment]
Agreed. Let's limit our dependencies to completed standards.

-- Mike


________________________________

**Comment by panva** [comment]
> @ilaril Thanks for your comment.
> 
> > I think that key ops stuff is not correct.
> 
> First of all, although the alg value is defined as the combination of KEM/KDF/AEAD, the alg parameter for COSE_Key is optional and I think the key_ops parameter is only for the KEM keys. Unless considering this way, I think we cannot keep the consistency with the previous definition of the use of key_ops below:
> 
> The section 6.3.1 (Direct ECDH) in RFC9053 says that the key_ops must be used as follows:
> 
> > If the "key_ops" field is present, it MUST include "derive key" or "derive bits" for the private key.
> > If the "key_ops" field is present, it MUST be empty for the public key.
> 
> In the case of the section 6.4 (ECDH with Key Wrap):
> 
> > If the "key_ops" field is present, it MUST include "derive key" or "derive bits" for the private key.
> > If the "key_ops" field is present, it MUST be empty for the public key.
> 
> In addition, unless defining key_ops in this way, the key cannot be available on the Web Cryptography API. It will be considered an invalid key (this is the case of JWK but we have to consier the consistency with JWK definition).
> 
> > Looks like empty key_ops makes the key unusable,
> 
> Actually, my implementation also has a bug in the processing of key_ops (I have to fix it as soon as possible), but are there any other examples that the keys that have empty key_ops cannot be used?
> 
> > and DeriveBits is very strange operation for decryption (since it is explicitly not for keys).
> 
> I agree that deriveBits is a strange thing.
> 
> However, since the Web Cryptography API used to implement a HPKE KEM operation is `deriveBits`, not `deriveKey`, I think we should limit the use of key_ops to `deriveBits`. In fact, the shared_secret generated as a result of KEM is not a key, just keying material.

I haven't undergone a WebCryptoAPI based implementation yet, but i will soon in the [jose](https://github.com/panva/jose) module. That being said, I believe that, considering the [current editor's draft of WebCryptoAPI](https://w3c.github.io/webcrypto) as well as use of [Secure Curves](https://wicg.github.io/webcrypto-secure-curves/)

- the JWK import operations don't limit the `alg` value of a JWK, which is good
- key_ops are empty for public keys
- key_ops are limited to `derive_bits` and/or `derive_key`

And so the definition in this PR is consistent with what WebCryptoAPI allows you to import/generate and the keys will be usable given this PR is in place AS-IS.

I will come back when I have more implementer's feedback under the current WebCryptoAPI state constraints.

**Comment by OR13** [comment]
@panva thanks for your review. 

I found that I needed to remove the unrecognized HPKE alg values before importing JWKs for use with dhkem.

I have a sorta side by side comparison of your library and my hacky attempt at HPKE using @dajiaji 's library at hpke.dev, but you should ignore it if you want to attempt a blind implementation based only on the specs.

**Comment by panva** [comment]
> I found that I needed to remove the unrecognized HPKE alg values before importing JWKs for use with dhkem.

Why? Which webcrypto implementation/runtime?

**Comment by panva** [comment]
Nevermind, answered myself - https://github.com/panva/jose/actions/runs/6833462247/job/18585240434, i'll open bug reports for both firefox and workerd and add WPTs for this behaviour. These bugs in implementations shouldn't drive the spec tho.

**Comment by dajiaji** [comment]
> Let's avoid signatures for this draft, there is work at cfrg that is still being worked out for hybrid signatures.

Needless to say, the JWK/COSE_Key expression for the PQ/T hybrid scheme should be defined as a separate spec independent of this COSE-HPKE spec :-)

**Comment by dajiaji** [comment]
@hannestschofenig Could you review and merge this PR? As far as I can see, there seems to be no opposition to the current status of the PR.

File: draft-ietf-cose-hpke.md [hunk 1]
```diff
@@ -253,6 +253,22 @@ The COSE_Encrypt MAY be tagged or untagged.
...
+
+The COSE_Key with the existing key types can be used to represent KEM private
+or public keys. When using a COSE_Key for COSE-HPKE, the following checks are made:
+
+* The "kty" field MUST be present, and it MUST be one of the key types for HPKE KEM.
+* If the "kty" field is "OKP" or "EC2", the "crv" field MUST be present
+  and it MUST be a curve for HPKE KEM.
+* If the "alg" field is present, the key type of its KEM MUST match the "kty" field.
+  And if the "kty" field is "OKP" or "EC2", the curve of the KEM MUST match the "crv" field.
+  The valid combinations of the "alg", "kty" and "crv" are shown in {{ciphersuite-kty-crv}}.
```
Comment (ilaril) [inline draft-ietf-cose-hpke.md:266]: The following might be a bit clearer (not repeating COSE requirements):

* The key MUST be convertible into key used by a HPKE KEM used by some algorithm.
  * Key with kty=EC2 and crv=P-256 can be converted into P256 key.
  * Key with kty=EC2 and crv=P-384 can be converted into P384 key.
  * Key with kty=EC2 and crv=P-521 can be converted into P521 key.
  * Key with kty=OKP and crv=X25519 can be converted into X25519 key.
  * Key with kty=OKP and crv=X448 can be converted into X448 key.
* If "alg" field is present, the key MUST be convertible into key for the HPKE KEM used by that algorithm.

... But probably requires some copyediting.

**Review by laurencelundblade — CHANGES_REQUESTED** [review]

**Comment by hannestschofenig** [comment]
I would like to merge this PR. Further refinements can then be made on top of it.

**Comment by laurencelundblade** [comment]
Let's be sure an capture the comments as issues if you merge. The PR doesn't correctly say how to use alg and curve yet.

File: draft-ietf-cose-hpke.md [hunk 2]
```diff
@@ -575,6 +616,53 @@ This example uses the following:
...
+### A KEM Private Key for HPKE-Base-P256-SHA256-AES128GCM
+
+~~~
+{
+    / kty = 'EC2' /
+    1: 2,
+    / kid = '01' /
+    2: h'3031',
+    / alg = HPKE-Base-P256-SHA256-AES128GCM (Assumed: 35) /
+    3: 35_0,
```
Comment (hannestschofenig) [inline draft-ietf-cose-hpke.md:653]: What does "35_0" mean? Why is this not just 35?

File: draft-ietf-cose-hpke.md [hunk 3]
```diff
@@ -575,6 +616,53 @@ This example uses the following:
...
+### A KEM Public Key for HPKE-Base-X25519-SHA256-CHACHA20POLY1305
+
+~~~
+{
+    / kty = 'OKP' /
+    1: 1,
+    / kid = '11' /
+    2: h'3131',
+    / alg = HPKE-Base-X25519-SHA256-CHACHA20POLY1305 (Assumed: 42) /
+    3: 42_0,
```
Comment (hannestschofenig) [inline draft-ietf-cose-hpke.md:678]: Shouldn't this be 42?

**Comment by dajiaji** [comment]
Thank you all for your reviews and comments. I will be able to get some time to update this PR this weekend. Please wait for a while.

File: draft-ietf-cose-hpke.md (hunk 2 above)
Comment (dajiaji) [reply draft-ietf-cose-hpke.md]: I've fixed the alg value.

File: draft-ietf-cose-hpke.md (hunk 3 above)
Comment (dajiaji) [reply draft-ietf-cose-hpke.md]: I've fixed the value.

File: draft-ietf-cose-hpke.md [hunk 4]
```diff
@@ -253,6 +253,22 @@ The COSE_Encrypt MAY be tagged or untagged.
...
+## Key Representation {#key-representation}
+
+The COSE_Key with the existing key types can be used to represent KEM private
+or public keys. When using a COSE_Key for COSE-HPKE, the following checks are made:
+
+* The "kty" field MUST be present, and it MUST be one of the key types for HPKE KEM.
+* If the "kty" field is "OKP" or "EC2", the "crv" field MUST be present
+  and it MUST be a curve for HPKE KEM.
+* If the "alg" field is present, the key type of its KEM MUST match the "kty" field.
+  And if the "kty" field is "OKP" or "EC2", the curve of the KEM MUST match the "crv" field.
```
[the remark this answers is not in the document]
Comment (dajiaji) [reply draft-ietf-cose-hpke.md]: @laurencelundblade Thanks for your review. I've added the sentence "it MUST be one of the supported COSE-HPKE "alg" values and ..." to the text. What do you think about that?

> If the "alg" field is present, **it MUST be one of the supported COSE-HPKE "alg" values and** the key type of its KEM MUST match the "kty" field.

**Pull request merged**